opencti-platform/opencti-graphql/src/database/engine.js

```diff
@@ -1336,25 +1338,43 @@ const elQueryBodyBuilder = async (context, user, options) => {
           const nestedElement = nested[nestIndex];
           const parentKey = validKeys.at(0);
           const { key: nestedKey, values: nestedValues, operator: nestedOperator = 'eq' } = nestedElement;
-          const nestedShould = [];
-          for (let i = 0; i < nestedValues.length; i += 1) {
-            const nestedFieldKey = `${parentKey}.${nestedKey}`;
-            const nestedSearchValues = nestedValues[i].toString();
-            if (nestedOperator === 'wildcard') {
-              nestedShould.push({ query_string: { query: `${nestedSearchValues}`, fields: [nestedFieldKey] } });
-            } else if (nestedOperator === 'not_eq') {
-              nestedMustNot.push({ match_phrase: { [nestedFieldKey]: nestedSearchValues } });
+          const nestedFieldKey = `${parentKey}.${nestedKey}`;
+          // Use term queries for keyword fields where applicable
+          // TODO: Get full list of keywordable fields but for now IDs are the most important use case
+          const keywordable = [ID_INTERNAL, ID_STANDARD, IDS_STIX, 'connections.internal_id'];
+          if (nestedOperator !== 'wildcard' && !(localFilterMode !== 'or' && isTermsSetUsable === false) && keywordable.indexOf(nestedFieldKey) !== -1) {
+            const keywordKey = `${nestedFieldKey}.keyword`;
+            let query;
+            if (localFilterMode === 'or') {
+              query = { terms: { [keywordKey]: nestedValues.map((v) => v.toString()) } };
             } else {
-              nestedShould.push({ match_phrase: { [nestedFieldKey]: nestedSearchValues } });
+              query = { terms_set: { [keywordKey]: { terms: nestedValues.map((v) => v.toString()), minimum_should_match: nestedValues.length } } };
```

Comment:
terms_set doesn't seem to work with `minimum_should_match` : 
`[terms_set] query does not support [minimum_should_match]`

According to the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/8.10/query-dsl-terms-set-query.html) : 

> The required number of matching terms is defined in the minimum_should_match_field or minimum_should_match_script parameter.

Reply:
Ah damn, did they get their release notes wrong or did I mess up the version check? (https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-8.10.0.html).

I will take that query offline and test it against local elasticsearch releases and see what I find.

Reply:
Oh I can see it has been added recently https://github.com/elastic/elasticsearch/pull/96082, but the documentation has not been updated it seems.

Reply:
Yeah they sort of ninja added it, took me a while to find it when I was hunting for a solution in the first place.

Reply:
I tested it on my local elasticsearch 8.9.1, not on 8.10, that's why I got the error. I'm going to test it on 8.10,

Reply:
So my version check is not working? (line 255)

Reply:
So the fix will only work on elasticsearch >= 8.10 ? otherwise we use the old match_phrase query ?

Reply:
> So my version check is not working? (line 255)

no no, I tested directly by quering my database, not with the code.

Reply:
So it won't work on opensearch either until I can work out what version they have added it in as well, but that can be a follow up. Assuming you guys are happy with the PR.

Reply:
I think OpenSearch adds the feature in `2.1` would you like me to check that and get that in the PR too?

Reply:
I wonder if it could be possible to use `minimum_should_match_script` which is already supported by older elasticsearch versions and opensearch : 
https://www.elastic.co/guide/en/elasticsearch/reference/8.10/query-dsl-terms-set-query.html#terms-set-query-script
https://opensearch.org/docs/2.0/opensearch/query-dsl/term/#terms-set

Reply:
Its doable (I think) but I don't know what the performance impact would be. I could try and implement it as a fallback for old elastics?

Reply:
I think we could use only "minimum_should_match_script" without having a test on elasticsearch/opensearch version, the code will be simpler.

Reply:
I would need to go and dig around the elastic source. But I dont like the idea of script evaluation each query which is why the feature was added. And would say the complexity is better to prevent performance impact. But it's not my product so am happy to do as directed.

Reply:
I am quite busy this week, but I will take this offline and get back to you with any findings.

Reply:
@SouadHadjiat based on this [ES Docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/scripts-and-search-speed.html#_improving_search_speed). I don't think its a good idea to use the script solution as a default. I still think versioning is a better solution but also acknowledge that the `and` use case is probably a rare one. Basically in our deployment the `too_many_clauses` is a huge issue for objects with many relations which are `or` based queries.